sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Redistribute.java

```diff
@@ -157,8 +168,16 @@ public boolean getAllowDuplicates() {
 
     @Override
     public PCollection<T> expand(PCollection<T> input) {
-      return input
-          .apply("Pair with random key", ParDo.of(new AssignShardFn<>(numBuckets)))
+      PCollection<KV<Integer, T>> sharded;
+      if (deterministicSharding) {
+        sharded =
+            input.apply(
+                "Pair with deterministic key",
+                ParDo.of(new AssignDeterministicShardFn<T>(numBuckets)));
```

Comment:
can you add a test to RedistributeTest for this?

Reply:
Do you have any tips for how to do this? I reviewed the existing RedistributeTest, but not sure how to set the offsets for elements or ProcessContext. If we can set it then we can verify repeated elements of the same offset get the same key.

Or are you thinking of just verifying that setting withDeterministicSharding using the new shard fn? Not sure how to verify yet, but can look into that one some more.

Reply:
Reworked to Kafka-specific redistribute and added testing.